### PR TITLE
Export WorkspaceWidget to workbench

### LINKED ISSUE #21442: Add MantidTreeWidget for workspaces to workbench
The mantid workbench requires a workspace widget just as mantid plot has. This is currently the [MantidTreeWidget](https://github.com/mantidproject/mantid/blob/master/qt/widgets/common/inc/MantidQtWidgets/Common/MantidTreeWidget.h).

This will be wrapped using sip to allow python access. However, in order to use it, several steps are required:
* [x] Remove dependency on MantidUI - Create a new MantidDisplayBase (`MantidTreeModel`) which only includes the functionality we require (current implementation is MantidPlot dependant)
* [x] Rewrite `QWorkspaceDockView`  to not be a `DockWidget` but rather a `QWidget`. Let parent class make decisions on docking. #21461 
* [x] `sip` wrapper
* [x] Minor test to ensure construction on workbench side - assume that tests for functionality are on the c++ side

### PR DESCRIPTION
**Pull request by ElliotAOram**
Exported the WorkspaceWidget to the workbench. Key features of this PR:
* Addition of MantidTreeModel (basic implementation on MantidUI - without the MantidPlot specifics)
* SIP wrapping for WorkspaceWidget
* Testcases and additions into workbench of SIP wrapper

**To test:**

* Code review

* Check the WorkspaceWidget runs as expected in the in the workbench
  * The widget appears in the bottom left
  * All the buttons at the top of the widget work as expected (same as MantidPlot)
  * All the *available* options from the context menu (right click on a workspace) work as expected.

Fixes #21442

**Release Notes** 
*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by ElliotAOram** [comment]
Before this is merged - the context menu needs to be constrained to only include the available options. 
I will have to think about the best solution for this as the WorkspaceWidget (where the context menu is populated) is shared between MantidPlot and the Workbench. The distinguishing feature between the two implementations (mantidplot vs. workbench) is `MantidUI.cpp` vs. `MantidTreeModel.cpp`. `MantidTreeModel` does not have access to the `WorkspaceWidget` directly (it just holds functions that are called by the `WorkspaceWidget`) not sure if this breaks MVP to give it access, and if it does have access if it can remove the options from the context menu. 
TLDR; investigation required for adding/remove context menu items.

File: qt/applications/workbench/workbench/plugins/workspacewidget.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+from mantidqt.widgets.workspacewidget.workspacetreewidget import WorkspaceTreeWidget
+from mantidqt.widgets.workspacewidget.mantidtreemodel import MantidTreeModel
+from qtpy.QtWidgets import QVBoxLayout
+
+# local package imports
+from workbench.plugins.base import PluginWidget
+
+
+class WorkspaceWidget(PluginWidget):
+    """Provides an in-process Jupyter Qt-based console"""
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plugins/workspacewidget.py]: A copy-and-paste error ?

File: qt/python/mantidqt/widgets/test/test_workspacetreewidget.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import unittest
+
+from qtpy.QtWidgets import QApplication
+
+from mantidqt.widgets.workspacewidget.mantidtreemodel import MantidTreeModel
+from mantidqt.widgets.workspacewidget.workspacetreewidget import WorkspaceTreeWidget
+
+QAPP = None
+
+
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/test/test_workspacetreewidget.py]: We have created a decorator to simplify tests that require a running QApplication so that we don't need to duplicate the `setupClass` stuff over and over: See [here](mantidproject/mantid/blob/5e0544dcdcf9e47149c67c2d0d3bfd4f102a6d5d/qt/applications/workbench/workbench/plugins/test/test_jupyterconsole.py#L30) for an example.

File: qt/python/mantidqt/widgets/workspacewidget/workspacetreewidget.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#   GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import (absolute_import)
+
+from mantidqt.utils.qt import import_qtlib
+
+
+WorkspaceTreeWidget = import_qtlib('_widgetscore', 'mantidqt.widgets.presenter', 'WorkspaceTreeWidget')
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/workspacewidget/workspacetreewidget.py]: I don't think the package name in the second argument is correct?

File: qt/widgets/common/src/MantidTreeModel.cpp [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+  Mantid::API::IAlgorithm_sptr alg;
+  try {
+    alg = Mantid::API::AlgorithmManager::Instance().create(
+        algName.toStdString(), version);
+  } catch (...) {
+    QString message = "Cannot create algorithm \"" + algName + "\"";
+    if (version != -1) {
+      message += " version " + QString::number(version);
+    }
+    QMessageBox::warning(nullptr, "MantidPlot", message);
```
Comment (martyngigg) [inline qt/widgets/common/src/MantidTreeModel.cpp]: I've recently discovered that if you leave the title argument of `QMessageBox` as an empty string then it inherits the current main window title. This is quite useful for us as we then don't need to specify it here.

File: qt/python/CMakeLists.txt [hunk 5]
```diff
@@ -51,6 +51,8 @@ if ( ENABLE_WORKBENCH )
 
   # ctest targets
   set ( PYTHON_TEST_FILES
+    mantidqt/test/algorithmselector_test.py
```
Comment (martyngigg) [inline qt/python/CMakeLists.txt]: I think there is a merge-related issue here. I think these 2 lines need to be removed.

**Comment by martyngigg** [comment]
Building locally I'm getting some warnings that I imagine the builders will show up too:
```
In file included from qt/widgets/common/qt4/inc/MantidQtWidgets/Common/moc_MantidTreeModel.cxx:9:0:
/media/data1/source/github/mantidproject/mantid/qt/widgets/common/inc/MantidQtWidgets/Common/MantidTreeModel.h:113:8: warning: ‘virtual void MantidQt::MantidWidgets::MantidTreeModel::executeAlgorithm(Mantid::API::IAlgorithm_sptr)’ can be marked override [-Wsuggest-override]
   void executeAlgorithm(Mantid::API::IAlgorithm_sptr alg);
        ^
/media/data1/source/github/mantidproject/mantid/qt/widgets/common/inc/MantidQtWidgets/Common/MantidTreeModel.h:124:8: warning: ‘virtual void MantidQt::MantidWidgets::MantidTreeModel::drawColorFillPlots(const QStringList&, GraphOptions::CurveType)’ can be marked override [-Wsuggest-override]
   void drawColorFillPlots(
        ^
[30/104] Building CXX object qt/widgets/common/CMakeFiles/MantidQtWidgetsCommonQt4.dir/src/MantidTreeModel.cpp.o
In file included from /media/data1/source/github/mantidproject/mantid/qt/widgets/common/src/MantidTreeModel.cpp:4:0:
/media/data1/source/github/mantidproject/mantid/qt/widgets/common/inc/MantidQtWidgets/Common/MantidTreeModel.h:113:8: warning: ‘virtual void MantidQt::MantidWidgets::MantidTreeModel::executeAlgorithm(Mantid::API::IAlgorithm_sptr)’ can be marked override [-Wsuggest-override]
   void executeAlgorithm(Mantid::API::IAlgorithm_sptr alg);
        ^
/media/data1/source/github/mantidproject/mantid/qt/widgets/common/inc/MantidQtWidgets/Common/MantidTreeModel.h:124:8: warning: ‘virtual void MantidQt::MantidWidgets::MantidTreeModel::drawColorFillPlots(const QStringList&, GraphOptions::CurveType)’ can be marked override [-Wsuggest-override]
   void drawColorFillPlots(
        ^
[34/104] Building CXX object qt/widgets/common/CMakeFiles/MantidQtWidgetsCommonQt4.dir/src/WorkspacePresenter/WorkspaceTreeWidgetSimple.cpp.o
/media/data1/source/github/mantidproject/mantid/qt/widgets/common/src/WorkspacePresenter/WorkspaceTreeWidgetSimple.cpp:24:58: warning: extra ‘;’ [-Wpedantic]
 WorkspaceTreeWidgetSimple::~WorkspaceTreeWidgetSimple(){};
                                                          ^
/media/data1/source/github/mantidproject/mantid/qt/widgets/common/src/WorkspacePresenter/WorkspaceTreeWidgetSimple.cpp:134:52: warning: unused parameter ‘matrixWS’ [-Wunused-parameter]
     const Mantid::API::MatrixWorkspace_const_sptr &matrixWS) const {
                                                    ^
[35/104] Building CXX object qt/widgets/common/CMakeFiles/MantidQtWidgetsCommonQt5.dir/src/WorkspacePresenter/WorkspaceTreeWidgetSimple.cpp.o
/media/data1/source/github/mantidproject/mantid/qt/widgets/common/src/WorkspacePresenter/WorkspaceTreeWidgetSimple.cpp:24:58: warning: extra ‘;’ [-Wpedantic]
 WorkspaceTreeWidgetSimple::~WorkspaceTreeWidgetSimple(){};
                                                          ^
/media/data1/source/github/mantidproject/mantid/qt/widgets/common/src/WorkspacePresenter/WorkspaceTreeWidgetSimple.cpp:134:52: warning: unused parameter ‘matrixWS’ [-Wunused-parameter]
     const Mantid::API::MatrixWorkspace_const_sptr &matrixWS) const {
```

Otherwise this looks good though.

**Review by martyngigg — APPROVED** [review]
This is a very cut down version of the workspace widget for demoing at the upcoming meeting. Further work will address some shortcomings in trying to share this code with the existing MantidPlot.
